.github/workflows/ca-ds-connection-test.yml

```diff
@@ -0,0 +1,120 @@
+
+      - name: Retrieve runner image
+        uses: actions/cache@v3
+        with:
+          key: pki-ca-runner-${{ inputs.os }}-${{ github.run_id }}
+          path: pki-runner.tar
+
+      - name: Load runner image
+        run: docker load --input pki-runner.tar
+
+      - name: Create network
+        run: docker network create example
+
+      - name: Set up DS container
+        run: |
+          tests/bin/ds-container-create.sh ds
+        env:
+          IMAGE: ${{ needs.init.outputs.db-image }}
+          COPR_REPO: ${{ needs.init.outputs.repo }}
+          HOSTNAME: ds.example.com
+          PASSWORD: Secret.123
+
+      - name: Connect DS container to network
+        run: docker network connect example ds --alias ds.example.com
+
+      - name: Set up PKI container
+        run: |
+          tests/bin/runner-init.sh pki
+        env:
+          HOSTNAME: pki.example.com
+
+      - name: Connect PKI container to network
+        run: docker network connect example pki --alias pki.example.com
+
+      - name: Install CA
+        run: |
+          docker exec pki pkispawn \
+              -f /usr/share/pki/server/examples/installation/ca.cfg \
+              -s CA \
+              -D pki_ds_hostname=ds.example.com \
+              -D pki_ds_ldap_port=3389 \
+              -D pki_cert_id_generator=random \
+              -D pki_request_id_generator=random \
+              -v
+
+      - name: Initialize PKI client
+        run: |
+          docker exec pki pki-server cert-export ca_signing --cert-file ca_signing.crt
+          docker exec pki pki client-cert-import ca_signing --ca-cert ca_signing.crt
+          docker exec pki pki info
+
+      - name: Create csr requesst
+        run: |
+          docker exec pki pki nss-cert-request --subject "CN=$HOSTNAME" --ext /usr/share/pki/server/certs/sslserver.conf  --csr sslserver.csr
+
+      - name: Test request enrollment
+        run: |
+          docker exec pki pki ca-cert-request-submit --profile caServerCert --csr-file sslserver.csr  | tee output
+          grep "Reason:" output | wc -l > actual
+          echo "0" > expected
+          diff expected actual
+          
+      - name: Stop the DS
+        run: |
+          docker stop ds
+          sleep 10
+          docker exec pki pki ca-cert-request-submit --profile caServerCert --csr-file sslserver.csr  | tee output
+          grep "Reason:" output | wc -l > actual
+          echo "1" > expected
+          diff expected actual
+
+      - name: Restart the DS
+        run: |
+          docker start ds
+          sleep 20
+          docker exec pki pki ca-cert-request-submit --profile caServerCert --csr-file sslserver.csr  | tee output
+          grep "Reason:" output | wc -l > actual
+          echo "0" > expected
+          diff expected actual
+
+      - name: Start without the DS
+        run: |
+          docker stop ds
+          sleep 20
+          docker exec pki pki-server restart
+          sleep 20
```

Comment:
Could we use this instead?
```
docker exec pki pki-server restart --wait
docker exec pki curl -s http://pki.example.com:8080/ca/admin/ca/getStatus
```

Reply:
The problem with the approach of minConns=0 is that the instance of `CMSEngine` is intialised during the start by the PKIWebListener and the initialisation requires to access the LDAP server so during the start-up you need to create the connections. If no connection are available the start get an error and the engine is shutdown. The initialisation is done with the method `start` which call serveral init methods asking the connection,  the first is `initPasswordStore` (actually this method does not use the LDAP connection factory but create its own connection, following init methods use the connection factory). I have tried to modify the code and delay the connection requests to the first access but this approach requires a lot of changes. Additionally, you have to add controls at each request and I do not think it is good so I give up to that approach.

Finally, maintain several connection open should be better for the performance. I have no idea of the typical workload of our customer but if plan to change this behaviour we should run some workload test before.

Reply:
> The problem with the approach of minConns=0 is that the instance of `CMSEngine` is intialised during the start by the PKIWebListener and the initialisation requires to access the LDAP server so during the start-up you need to create the connections. If no connection are available the start get an error and the engine is shutdown.

Yeah, I think this is actually the problem that we want to solve. I don't think PKI server actually needs to do any database operations during startup, so it's not necessary to create connections at that time. While creating connections at startup could improve performance, that's not always the main priority in all cases.

> The initialisation is done with the method `start` which call serveral init methods asking the connection, the first is `initPasswordStore` (actually this method does not use the LDAP connection factory but create its own connection, following init methods use the connection factory). I have tried to modify the code and delay the connection requests to the first access but this approach requires a lot of changes.

I warned you before that it might not be a trivial change :grimacing: 

> Additionally, you have to add controls at each request and I do not think it is good so I give up to that approach.

Which controls are you referring to? We should consider this as a connection pool. The caller should not be aware about how many database connections the pool has at any time.

> Finally, maintain several connection open should be better for the performance. I have no idea of the typical workload of our customer but if plan to change this behaviour we should run some workload test before.

People might have different priorities. Keeping connections open (and retrying to open connections) could cost more in a cloud environment. The admin should be able to adjust the `minConns` according to their needs. If they need performance, they can use a higher `minConns`. If they want to use the minimum resources they should be able to use `minConns=0` but the server should still be working.

For comparison, DBCP uses `initialSize=0` and `minIdle=0` by default:
https://commons.apache.org/proper/commons-dbcp/configuration.html

Reply:
I have not analysed in details all the initialisation steps. I will try to remove all the LDAP dependent initialisation from the start-up. If this is possible I will go with your plan so we can have minConns=0 (and eventually other parameter to control how the pool works).

If this is not possible then we should complete the initialisation when a request is performed and I am not sure it is good (the control I was referring is this, check if it is initialised and eventually do it). I will work on this.

I put on draft for the moment until the new version is ready.